influxdb3_write/src/last_cache.rs

```diff
@@ -0,0 +1,498 @@
+        Self {
+            cache_map: Default::default(),
+        }
+    }
+
+    /// Output the records for a given cache as arrow [`RecordBatch`]es
+    #[cfg(test)]
+    pub(crate) fn get_cache_record_batches<D, T>(
+        &self,
+        db_name: D,
+        tbl_name: T,
+    ) -> Option<Result<RecordBatch, ArrowError>>
+    where
+        D: AsRef<str>,
+        T: AsRef<str>,
+    {
+        self.cache_map
+            .read()
+            .get(db_name.as_ref())
+            .and_then(|db| db.get(tbl_name.as_ref()))
+            .map(|lc| lc.read().to_record_batch())
+    }
+
+    /// Create a new entry in the last cache for a given database and table, along with the given
+    /// parameters.
+    pub(crate) fn create_cache<D, T>(
+        &self,
+        db_name: D,
+        tbl_name: T,
+        count: usize,
+        key_columns: impl IntoIterator<Item: Into<String>>,
+        schema: SchemaRef,
+    ) -> Result<(), Error>
+    where
+        D: Into<String>,
+        T: Into<String>,
+    {
+        let db_name = db_name.into();
+        let tbl_name = tbl_name.into();
+        if self
+            .cache_map
+            .read()
+            .get(&db_name)
+            .is_some_and(|db| db.contains_key(&tbl_name))
+        {
+            return Err(Error::CacheAlreadyExists);
+        }
+        let last_cache = RwLock::new(LastCache::new(count, key_columns, schema)?);
+        self.cache_map
+            .write()
+            .entry(db_name)
+            .or_default()
+            .insert(tbl_name, last_cache);
+        Ok(())
+    }
+
+    /// Write a batch from the buffer into the cache by iterating over its database and table batches
+    /// to find entries that belong in the cache.
+    ///
+    /// Only if rows are newer than the latest entry in the cache will they be entered.
+    pub(crate) fn write_batch_to_cache(&self, write_batch: &WriteBatch) {
+        for (db_name, db_batch) in &write_batch.database_batches {
+            for (tbl_name, tbl_batch) in &db_batch.table_batches {
+                if let Some(db) = self.cache_map.read().get(db_name.as_str()) {
+                    for row in &tbl_batch.rows {
+                        if db
+                            .get(tbl_name)
+                            .is_some_and(|t| row.time > t.read().last_time.timestamp_nanos())
+                        {
+                            db.get(tbl_name).unwrap().write().push(row);
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Stores the last N values, as configured, for a given table in a database
+pub(crate) struct LastCache {
+    // TODO: not sure if this is needed, given the individual columns track their size
+    _count: LastCacheSize,
+    // TODO: use for filter predicates
+    _key_columns: HashSet<String>,
+    schema: SchemaRef,
+    // use an IndexMap to preserve insertion order:
+    cache: IndexMap<String, CacheColumn>,
```

Comment:
It's not clear to me why this is being used here. What's the insertion order being preserved here?

Reply:
Since I insert columns into the cache using the ordering of fields in the `schema` ([here](https://github.com/influxdata/influxdb/pull/25109/files#diff-6740f1021d631fa570625b9b27f8b4692fd0777c50eabcc435cda7793d0da049R155-R159)), then when producing a record batch out of the cache ([here](https://github.com/influxdata/influxdb/pull/25109/files#diff-6740f1021d631fa570625b9b27f8b4692fd0777c50eabcc435cda7793d0da049R184-R193)), `IndexMap` allows to iterate over the map directly while producing the correct order of columns for the schema.

Reply:
The links in that comment look to be out-of-date. But I have added docs/comments in the code to help explain this.

Reply:
@pauldix - while working on https://github.com/influxdata/influxdb/pull/25125 to enable caches that add new fields, I have found that the insertion order guarantee falls apart in scenarios where writes come in with different fields/orders for different key values. So, we can't rely on that, and I will be removing the comments about insertion order.

I still like the use of `IndexMap` because it gives fast iteration over keys, as well as fast lookups (see [here](https://github.com/indexmap-rs/indexmap?tab=readme-ov-file#performance)), but if we want to avoid the dependency, or just optimize for lookup speed, then we could use a `HashMap` here.